tests/RobotFramework/tests/tedge_agent/workflows/dynamic_workflow_reloading.robot

```diff
@@ -0,0 +1,250 @@
+    Execute Command
+    ...    tedge mqtt pub --retain te/device/main///cmd/user-command/dyn-test-2 '{"status":"init"}'
+    Should Have MQTT Messages
+    ...    te/device/main///cmd/user-command/dyn-test-2
+    ...    message_pattern=.*successful.*
+    ${workflow_log}    Execute Command    cat /var/log/tedge/agent/workflow-user-command-dyn-test-2.log
+    Should Contain
+    ...    ${workflow_log}
+    ...    item="@version":"1370727b2fcd269c91546e36651b9c727897562a5d3cc8e861a1e35f09ec82a6"
+    Should Contain    ${workflow_log}    item="user-command":"second-version"
+
+Remove User-Defined Operation
+    ThinEdgeIO.File Should Exist    /etc/tedge/operations/user-command.toml
+    ${timestamp}    Get Unix Timestamp
+    Execute Command    rm /etc/tedge/operations/user-command.toml
+    ${capability}    Should Have MQTT Messages    te/device/main///cmd/user-command    date_from=${timestamp}
+    Should Be Empty    ${capability[0]}
+
+Updating A Workflow Twice Before Using It
+    ThinEdgeIO.File Should Not Exist    /etc/tedge/operations/user-command.toml
+    ThinEdgeIO.Transfer To Device    ${CURDIR}/user-command-v1.toml    /etc/tedge/operations/user-command.toml
+    ${capability}    Should Have MQTT Messages    te/device/main///cmd/user-command
+    Should Be Equal    ${capability[0]}    {}
+    ThinEdgeIO.Transfer To Device    ${CURDIR}/user-command-v2.toml    /etc/tedge/operations/user-command.toml
+    ${capability}    Should Have MQTT Messages    te/device/main///cmd/user-command
+    Should Be Equal    ${capability[0]}    {}
+    Execute Command
+    ...    tedge mqtt pub --retain te/device/main///cmd/user-command/dyn-test-3 '{"status":"init"}'
+    Should Have MQTT Messages
+    ...    te/device/main///cmd/user-command/dyn-test-3
+    ...    message_pattern=.*successful.*
+    ${workflow_log}    Execute Command    cat /var/log/tedge/agent/workflow-user-command-dyn-test-3.log
+    Should Contain
+    ...    ${workflow_log}
+    ...    item="@version":"1370727b2fcd269c91546e36651b9c727897562a5d3cc8e861a1e35f09ec82a6"
+    Should Contain    ${workflow_log}    item="user-command":"second-version"
+
+Override Builtin Operation
+    ThinEdgeIO.File Should Not Exist    /etc/tedge/operations/software_list.toml
+    ThinEdgeIO.Transfer To Device    ${CURDIR}/software_list.toml    /etc/tedge/operations/software_list.toml
+    Execute Command
+    ...    tedge mqtt pub --retain te/device/main///cmd/software_list/dyn-test-4 '{"status":"init"}'
+    Should Have MQTT Messages
+    ...    te/device/main///cmd/software_list/dyn-test-4
+    ...    message_pattern=.*successful.*
+    ${workflow_log}    Execute Command    cat /var/log/tedge/agent/workflow-software_list-dyn-test-4.log
+    Should Contain
+    ...    ${workflow_log}
+    ...    item="@version":"76e9afe834b4a7cadc9029670ba76745fcda73784f9e78c09f0c0416f7f58ad2"
+
+Recover Builtin Operation
+    ThinEdgeIO.File Should Exist    /etc/tedge/operations/software_list.toml
```

Comment:
suggestion: the test fails if run by itself and not part of the test suite, because this file is created by the previous test case. Could we instead use Transfer to Device everywhere so that there are no dependencies between the test cases? I'd expect Transfer to Device to overwrite the file if it already exists, so it should be okay to use that?

Reply:
I have a mix feeling here. On one side, you are correct, it would be handy to have independent tests. But, on the other side, this test suite represents well a scenario where a user creates and iterate updating a workflow file.

Concretely, replacing`File Should Exist` assertion by a `Transfer to Device` command would lead to a *different* test while running the suite vs the isolated test. Indeed, in the suite case, one checks that a user can update a workflow while, in the isolated case, one checks that the user can create a workflow (i.e. `Update User-Defined Operation` doing the same test as `Create User-Defined Operation`).